helper/winsvc/mock_windows_service.go

```diff
@@ -0,0 +1,297 @@
+	return &MockWindowsService{t: t}
+}
+
+type MockWindowsServiceManager struct {
+	isServiceRegistereds [][3]any
+	getServices          [][3]any
+	createServices       [][5]any
+	t                    *testing.T
+}
+
+func (m *MockWindowsServiceManager) ExpectIsServiceRegistered(name string, result bool, err error) {
+	m.isServiceRegistereds = append(m.isServiceRegistereds, [3]any{name, result, err})
+}
+
+func (m *MockWindowsServiceManager) ExpectGetService(name string, result WindowsService, err error) {
+	m.getServices = append(m.getServices, [3]any{name, result, err})
+}
+
+func (m *MockWindowsServiceManager) ExpectCreateService(name, binaryPath string, config WindowsServiceConfiguration, result WindowsService, err error) {
+	m.createServices = append(m.createServices, [5]any{name, binaryPath, config, result, err})
+}
+
+func (m *MockWindowsServiceManager) IsServiceRegistered(name string) (bool, error) {
+	m.t.Helper()
+
+	must.SliceNotEmpty(m.t, m.isServiceRegistereds,
+		must.Sprint("Unexpected call to IsServiceRegistered"))
+	call := m.isServiceRegistereds[0]
+	m.isServiceRegistereds = m.isServiceRegistereds[1:]
+	must.Eq(m.t, call[0].(string), name,
+		must.Sprint("IsServiceRegistered received incorrect argument"))
+
+	if call[2] == nil {
+		return call[1].(bool), nil
+	}
+
+	return call[1].(bool), call[2].(error)
+}
+
+func (m *MockWindowsServiceManager) GetService(name string) (WindowsService, error) {
+	m.t.Helper()
+
+	must.SliceNotEmpty(m.t, m.getServices,
+		must.Sprint("Unexpected call to GetService"))
+	call := m.getServices[0]
+	m.getServices = m.getServices[1:]
+	must.Eq(m.t, call[0].(string), name,
+		must.Sprint("GetService received incorrect argument"))
+
+	if call[2] == nil {
+		return call[1].(WindowsService), nil
+	}
+
+	return nil, call[2].(error)
+}
+
+func (m *MockWindowsServiceManager) CreateService(name, binaryPath string, config WindowsServiceConfiguration) (WindowsService, error) {
+	m.t.Helper()
+
+	must.SliceNotEmpty(m.t, m.createServices,
+		must.Sprint("Unexpected call to CreateService"))
+	call := m.createServices[0]
+	m.createServices = m.createServices[1:]
+	must.Eq(m.t, call[0].(string), name,
+		must.Sprint("CreateService received incorrect argument"))
+	must.StrContains(m.t, binaryPath, call[1].(string),
+		must.Sprint("CreateService received incorrect argument"))
+
+	cfg := call[2].(WindowsServiceConfiguration)
+	if !reflect.ValueOf(cfg).IsZero() {
+		must.Eq(m.t, call[2].(WindowsServiceConfiguration), config,
+			must.Sprint("CreateService received incorrect argument"))
+	}
+
+	if call[4] == nil {
+		return call[3].(WindowsService), nil
+	}
+
+	return nil, call[4].(error)
+}
+
+func (m *MockWindowsServiceManager) Close() error { return nil }
+
+func (m *MockWindowsServiceManager) AssertExpectations() {
+	must.SliceEmpty(m.t, m.isServiceRegistereds,
+		must.Sprintf("IsServiceRegistered expecting %d more invocations", len(m.isServiceRegistereds)))
```

Comment:
it's not intuitive to me that "`AssertExpectations`" means "assert that nothing has been called"

also the message seems to indicate that the slices should _not_ be empty? I suppose I'm pretty confused by this method in general.

Reply:
This is just verifying that all expected calls to the mock instance were actually called. When the functions are invoked, they remove the entry so after the test has finished, all of them should be empty. So it's just asserting that _everything_ has been called.

The checks are checking that the slices are empty. If they are not empty, then more calls were expected, and the annotated error message will include which function was still expecting to be called, and how many more times were expected.